[WIP] Compare performance

palimondo opened the pull request:
<!-- What's in this pull request? -->
This PR fixes reporting issues when comparing benchmark results.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves: 
* [SR-4590](https://bugs.swift.org/browse/SR-4590) `compare_perf_tests.py` fails when new benchmarks are added
* [SR-4597](https://bugs.swift.org/browse/SR-4597) Benchmark results have wrong MEAN, MEDIAN and SD
* [SR-4598](https://bugs.swift.org/browse/SR-4598) Add option to run subset of benchmarks matching a prefix
* [SR-4659](https://bugs.swift.org/browse/SR-4659) Benchmark logs should be tied to tested tree version
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

palimondo committed:
Fix: SR-4597 Benchmark results have wrong MEAN, MEDIAN and SD

palimondo committed:
Fix SR-4598 Add option to run subset of benchmarks matching a prefix

palimondo committed:
Fix SR-4590 compare_perf_tests.py fails when new benchmarks are added

palimondo commented:
@lplarson Please review.

shahmishal commented:
@swift-ci please benchmark

palimondo commented:
I’m not done here yet, still working on [SR-4600](https://bugs.swift.org/browse/SR-4600) and [SR-4601](https://bugs.swift.org/browse/SR-4601), but I have requested the review for the changes do far - mainly the fix for `Benchmark_Driver` to correctly report MEAN and SD.

gottesmm commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -81,42 +81,14 @@ def submit_to_lnt(data, url):
-
-    return avg_test_output
-
+    test_output_raw = subprocess.check_output(
```
So just to make sure I am understanding this correctly, you are instrumenting the max resident size of the process tree rooted at the Driver given a specific test.

I think that will give you the max, but it won't allow you to calculate other statistics over the samples. I haven't thought about the implications of this... @lplarson your thoughts on this?

palimondo replied:
Yes, as I understand this, this just gets the max memory used by the given test. This will not be a statistical sample, just a single value. All other stats: MIN, MAX, MEAN, SD, MEDIAN are correctly  reported by `Benchmark_O` & co. and were — before this patch — dirscarded or corrupted by this script.

lplarson replied:
We want to be able to take an average of the peak memory usages.

palimondo replied:
We first need to get the basics working at all. The script as is, does not report average memory use at all, but a sum of memory used in all runs — results depend on `num-iterations`. The `MAX_RSS` is not being reported by the `compare_perf_tests.py` either.

I have compared two runs (`num-iterations` 3 and 20) with the fixed script on my machine and the difference in `MAX_RSS` between the runs for all tests is in 0-1% range, with the exception of following: 

Testname | Mem Diff 
----------|--------------
Array2D | 2 % 
ArrayAppendSequence | 2 %
ArrayAppend | 3%
ArrayAppendStrings | 4 %
ArrayPlusEqualSingleElementCollection | 2 %
ArraySubscript | 9 %
DictionaryBridge | 90 %
MonteCarloE | 3 %
ObjectiveCBridgeStubFromNSStringRef | 9 %
ObjectiveCBridgeStubToNSStringRef | 7 %
ObjectiveCBridgeToNSDictionary | 84 %
ObjectiveCBridgeToNSSet | 84 %
ObserverClosure | 2 %
ObserverPartiallyAppliedMethod | 3 %

It looks like there are some issues (leaks ?) with ObjC Bridging (DictionaryBridge, ObjectiveCBridgeToNSDictionary, ObjectiveCBridgeToNSSet). Let me emphasize that I was comparing run with 3 vs 20 samples, done by `Benchmark_O`. 

I don’t see what could be gained by computing average memory used over multiple runs executed by `Benchmark_Driver` — are we really expecting Swift to have varying memory consumption between runs?!? That approach resulted in the need to redo all the statistics `Benchmark_O` already does in `Benchmark_Driver`- *none of which* was implemented correctly.

palimondo replied:
@lplarson @gottesmm Can I please get a response to the argument made in the above comment?

lplarson replied:
Averaging is useful here for the same reason it's useful in measuring benchmark runtime. These systems always have noise and averaging helps deal with it. We moved to using Benchmark_Driver specifically for the reason of averaging over memory usage.

lplarson replied:
"The script as is, does not report average memory use at all, but a sum of memory used in all runs" Where do you see this?

palimondo replied:
Re: "script doesn’t report average but sum” - I cannot reproduce what I saw. It definitely wasn’t caused by this script, but I have a few logs where the MAX_RSS dropped from about 10 MB (for small tests) to 3,3 MB. This lead me to believe the change was caused by my modification of the script (I was testing with `num-iterations=3`). Now the runs are back to 10 MB. I have no idea what is happening here, will investigate further.

palimondo replied:
I’m taking this conversation to the top, being hidden under comments to outdated file is not helping.

gottesmm commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -81,42 +81,14 @@ def submit_to_lnt(data, url):
-    return avg_test_output
-
+    test_output_raw = subprocess.check_output(
+        ['time', '-lp', driver_path, test, '--num-samples='+str(num_samples)],
```
Can you use a string format rather than str? i.e.: '--num-samples={}'.format(num_samples) or using '%s'

gottesmm commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -90,9 +90,17 @@ def instrument_test(driver_path, test, num_samples):
     """Return a list of available performance tests"""
-    return subprocess.check_output([driver_path, '--list']).split()[2:]
+    tests = subprocess.check_output([driver_path, '--list']).split()[2:]
+    if args.filter:
```
Can you use early returns here to eliminate the elif. You do not have any code of consequence after the condition. This is communicated to the reader of the code in a more succint way by using early returns. i.e.:

```
if args.filter:
    prefix = args.filter
    return filter(lambda name: name.startswith(prefix), tests)

if not args.benchmark:
    return tests

benchmarks = set(args.benchmark)
return sorted(list(set(tests).intersection(benchmarks)))
```

Notice how just by reading the code the reader is able to know the amount of work left to be returned without having to read the whole function.

gottesmm reviewed (commented):
Some quick comments. I don't completely understand what you are trying to do with the 2nd patch and I have not read the first. I am going to look at this again later.

lplarson commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -81,46 +81,27 @@ def submit_to_lnt(data, url):
-
-def get_tests(driver_path):
+    test_output_raw = subprocess.check_output(
+        ['time', '-lp', driver_path, test, '--num-samples=' +
```
Using --num-samples and getting the peak memory usage of a single invocation doesn't allow us to average the memory usage for each sample.

lplarson replied:
I'd like to be able to average over peak memory usages.

palimondo committed:
Addressed python style issue raised during review

palimondo committed:
Refactored Markdown formatting

palimondo commented:
I have a question about output from `compute_perf_test.py`: currently it writes output to file specified in `--output` argument in the format specified in `--format` argument, but *always writes to `stdout`*!  Either markdown (for `markdown` or no specified format)  or `git` format (for `html` or `git`).

I propose that it always writes *only in specified format* and that it writes either to `stdout` OR the file,  when `--output` is specified.

palimondo commented:
Can I please get a guidance about the script outputs I suggest altering in the above comment?
I’d like to change `compare_perf_test.py` to behave like standard common line utility. At the moment it does two outputs simultaneously: to`stdout` in one format, to output file - if specified - in another format. Is that a desired behavior that something else depends on?
@lplarson @gottesmm @shahmishal

palimondo committed:
Refactoring: Condensed HTML formatting, extracted Markdown formatting and list computations

palimondo commented:
OMG, I need help from git expert - I messed something up here. I wasn’t able to push my recent commits. It said I needed to git pull first, I played with git rebase, reset, clean and co… and now I’ve somehow pulled in commit from @atrick about “CopyForwarding” and have one of mine duplicated. Agh. Help (e-mail, twitter)?

lplarson commented:
@palimondo Can you rebuild your branch and force push?

palimondo commented:
I’ll try to fix my branch and force push. 

Here I’d like to continue our discussion from the hidden comments on outdated `Benchmark_Driver`.

@lplarson: Maybe you could explain more about the reasoning that lead to iterating in `Benchmark_Driver`?

When measuring runtime performance on multitasking system, we have to take into account the system’s weather: other processes competing for the CPU. We are taking multiple samples and trying to statistically approximate a “true” runtime. Instead of measuring how would our benchmark perform if it was the only thing running — this is never the case — we settle for a consistently reproducible result. We should be discarding extreme values and work with the mean.

In case of measuring memory usage, the system's weather is much less severe. There certainly is variation due to interaction between allocator and OS that depends on the available memory, but this is small compared to measuring runtime. Other processes do not directly impact our own memory consumption.

We are measuring the memory usage by instrumenting the benchmark run with `time -lp` command that adds the following epilogue to the test run:

```
#,TEST,SAMPLES,MIN(us),MAX(us),MEAN(us),SD(us),MEDIAN(us)
1,AngryPhonebook,2,10799,12206,11502,994,12206

Totals,1,10799,12206,11502,0,0
real         2.09
user         1.93
sys          0.01
   3469312  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
       905  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
       514  involuntary context switches
```

From this we extract the Maximum [resident set size](https://en.wikipedia.org/wiki/Resident_set_size). The granularity of this value is always a multiple of page size - **4098** on macOS. That is the smallest amount of memory the operating system deals with - the smaller allocations amounts used internally by the program are handled by the allocator (malloc) and that is part of the Swift runtime. In theory it is possible to measure memory usage on finer level by using a custom allocator to track all the details, but this is what we are discussing here.

When we look at the vast amount of tests in the Swift benchmark suite, as long as they are pure Swift, they are very stable memory wise. Running `Benchmark_Driver run -o O -i 3 AngryPhonebook` 4 times gives: 

| Benchmark_Driver iterates | Samples | MAX_RSS(B) | Deviation | CV   | 
|---------------------------|---------|------------|-----------|------| 
| AngryPhonebook            | 3       | 10 519 893 | -683      | -0 % | 
| AngryPhonebook            | 3       | 10 523 989 | -4 779    | -0 % | 
| AngryPhonebook            | 3       | 10 517 163 | 2 048     | 0 %  | 
| AngryPhonebook            | 3       | 10 515 797 | 3 414     | 0 %  | 
| **Average**                   |         | **10 519 211** | **3 612**     |  | 

Each **MAX_RSS** value is the mean of 3 samples taken by the current version of `Benchmark_Driver`. The last row is mean of the means from the 4 runs (i.e. average from 12 samples). **Deviation** column shows the difference between **MAX_RSS** and the **Average RSS**.  The last value in **Deviation** column is the standard deviation (**SD**), **CV** is  [coefficient of variation](https://en.wikipedia.org/wiki/Coefficient_of_variation) defined as the ratio of the standard deviation to the mean. 

Raising the sample size to 20 buys us very little:

| Benchmark_Driver iterates | Samples | MAX_RSS(B) | Deviation | CV   | 
|---------------------------|---------|------------|-----------|------| 
| AngryPhonebook            | 20      | 10 520 781 | -1 434    | -0 % | 
| AngryPhonebook            | 20      | 10 514 637 | 4 710     | 0 %  | 
| AngryPhonebook            | 20      | 10 522 829 | -3 482    | -0 % | 
| AngryPhonebook            | 20      | 10 519 142 | 205       | 0 %  | 
| Average                   |         | 10 519 347 | 3 484     | 0 %  | 


Notice how the variation between **MAX_RSS** values reported from different runs did not decrease even when we increased the sample size. This is because we are now doing statistical analysis of the noise that is below the precision of our measurement. The point to take home is that **CV** is hugging 0%.

If we ditched averaging of RSS and left the computation of other statistics to `Benchmark_O` and co. as I suggest in my patch, the result looks like this:

|`Benchmark_0`<br/>`--num-samples`| Samples | MAX_RSS(B)   | Deviation | CV   | #Pages  | Delta | 
|---------------------------|---------|--------------|-----------|------|---------|-------| 
| AngryPhonebook            | 20      | 11 079 680   | 2 048     | 0 %  | 2 705   | 0     | 
| AngryPhonebook            | 20      | 11 079 680   | 2 048     | 0 %  | 2 705   | 0     | 
| AngryPhonebook            | 20      | 11 087 872   | -6 144    | -0 % | 2 707   | 2     | 
| AngryPhonebook            | 20      | 11 079 680   | 2 048     | 0 %  | 2 705   | 0     | 
| **Average**                 |         | **11 081 728** | **4 096**   |      | **2 705** |       | 

At first glance the **SD**looks worse, but to clarify I’ve added 2 columns: the number of memory pages allocated (MAX_RSS / 4096) and the delta between the minimal number of pages (#Pages, last row) and the current one.

Notice the different sizes reported between the two versions of `Benchmark_Driver` — about 130 pages or 500 kB — they stay same even for `--num-samples=1`, so I guess this is all due to driver processing the `--num-samples` argument from command line.

My argument is that while the CV stays low, we are good to stake just one measurement and report memory regressions (not done currently at all) based on the % change between them, say above 5%.

This takes me to further analyze tests that had high CV — I’ve referred to that as **Mem Diff** in the table of suspicious test in my previous comment. I believe they are all a symptom of a bug: memory leaks in the bridging implementation. This is actually hidden by current, MAX_RSS averaging, implementation of `BenchmarkDriver`. When measured by my modified driver I see this:

|`Benchmark_0`<br/>`--num-samples`| S. | MAX_RSS(B)     | Deviation     | CV   | #Pages    | Delta |      | 
|---------------------------|---------|----------------|---------------|------|-----------|-------|------| 
| DictionaryBridge          | 3       | 36 966 400     | 4 481 024     | 12 % | 9 025     | 0     | 0 %  | 
| DictionaryBridge          | 3       | 42 110 976     | -663 552      | -2 % | 10 281    | 1 256 | 12 % | 
| DictionaryBridge          | 3       | 44 720 128     | -3 272 704    | -7 % | 10 918    | 1 893 | 17 % | 
| DictionaryBridge          | 3       | 41 992 192     | -544 768      | -1 % | 10 252    | 1 227 | 12 % | 
| **Average**               |         | **41 447 424** | **3 241 771** |      | **9 025** |       |      | 


| `Benchmark_0`<br/>`--num-samples` | S. | MAX_RSS(B)      | Deviation      | CV   | #Pages     | Delta  |      | 
|---------------------------|---------|-----------------|----------------|------|------------|--------|------| 
| DictionaryBridge          | 20      | 179 634 176     | 33 456 128     | 19 % | 43 856     | 0      | 0 %  | 
| DictionaryBridge          | 20      | 203 837 440     | 9 252 864      | 5 %  | 49 765     | 5 909  | 12 % | 
| DictionaryBridge          | 20      | 235 429 888     | -22 339 584    | -9 % | 57 478     | 13 622 | 24 % | 
| DictionaryBridge          | 20      | 233 459 712     | -20 369 408    | -9 % | 56 997     | 13 141 | 23 % | 
| **Average**               |         | **213 090 304** | **26 576 291** |      | **43 856** |        |      | 

Going from 3 to 20 samples increases memory consumption 5 times! Based on that I’m lumping **ObjectiveCBridgeToNSDictionary** and **ObjectiveCBridgeToNSSet** together as the same bug.

The other volatile test are much less severe (CV < 10%) and have in common intense use of Array. I’ve looked in detail at **MonteCarloE**. It performs a lot of random number computations, similar to MonteCarloPi, but in contrast to that it builds a huge [Bool]. 

| `Benchmark_0 --num-samples` | Samples | Average MAX_RSS(B) | 
|-------------------------------|---------|--------------------| 
| MonteCarloE                   | 1       | 22 128 933         | 
| MonteCarloE                   | 3       | 26 537 984         | 
| MonteCarloE                   | 20      | 32 735 232         | 

This looks like a much less severe, but still some kind of memory leak in the implementation of Array. Given that the `ContinuousArray` is recommended for use with pure Swift code, that doesn’t need to interoperate with ObjC, I’m guessing that `Array` always carries some kind of baggage for bridging support that could be similar to the above issues with Set and Dictionary bridging???

**ObserverPartiallyAppliedMethod** had a up to 5% CV with 3 iterations but stabilized to 0-1% with 20 iterations, the memory increase maybe worth investigating?

| `Benchmark_0 --num-samples`      | Samples | Average MAX_RSS(B) | 
|--------------------------------|---------|--------------------| 
| ObserverPartiallyAppliedMethod | 3       | 83 765 248         | 
| ObserverPartiallyAppliedMethod | 20      | 86 861 824         | 

All in all, I believe that unstable memory usage in benchmarks is a sign of bugs in underlying implementation - not a fact of nature that we need to deal with by averaging. This approach allowed for the above bugs to hide in plain sight for who-knows-how-long. 

Or am I really misunderstanding some fundamental issue that lead you to compute average MAX_RSS in the `Benchmark_Driver`?

palimondo commented:
I have fixed my branch and force pushed. 

@lplarson @shahmishal @gottesmm Can I get some answers here? I’d like to continue the work on this, but I’d hate to throw it away if you don’t like where I’d like to take it. See the 4-5 comments above.

palimondo committed:
Fix SR-4659 Benchmark logs should be tied to tested tree version

palimondo committed:
Benchmark_Driver - Report totals for all columns

palimondo commented:
FWIW, I’ve filed 
* [SR-4666](https://bugs.swift.org/browse/SR-4666) Dictionary bridge is leaking memory
* [SR-4667](https://bugs.swift.org/browse/SR-4667) Array is probably leaking memory
* [SR-4668](https://bugs.swift.org/browse/SR-4668) NSStringRef is probably leaking memory
to track the issue of the most memory unstable benchmarks.

palimondo commented:
Note about the current `compare_perf_test.py`script's always-present output to `stdout`:<br/>when the `--format` is **markdown**, it outputs it; otherwise it prints out in **`PAIN_DETAIL`** format — that is markdown with "-" and "|" replaced by " ".  Is that a desired and behavior? (If so, could we at least  _always_ output to console in the same format?)

atrick commented:
No idea what you're actually proposing. The original motivation was to paste the output into a git message or copy paste some lines into emacs org mode and `org-table-convert-region`. I think the `--output` option is for integration with CI. I don't use it and don't really understand why it would be tied to the format.

I'd be fine with just a "markdown" format by default, as it is today, and "text" format that has the markup and table characters stripped away.

atrick commented:
BTW: what we *really* need is a Benchmark_Driver --rerun=N feature: https://bugs.swift.org/browse/SR-4669

palimondo commented:
@atrick I'm proposing that the script always outputs in the format specified by `--format` attribute and that it either writes to stdout or to file if `--output` is specified.

atrick commented:
@palimondo That sounds completely sane. Thanks. As long as it's ok with @shahmishal.

palimondo committed:
Refactoring compare_perf_test.py complete

Created ReportFormatter. The script still produces same output as original legacy version.

gottesmm commented:
@palimondo Can you group the commits into different PRs. I have time to review today in a little bit.

palimondo commented:
@gottesmm I can, if you tell me what you want… Wrote you an e-mail.

gottesmm commented:
@palimondo I responded via email. Did I answer your question?

palimondo committed:
Reproduce legacy behavior for --change-only

Also: use single quote strings

lplarson commented:
@palimondo Thinking about this more, I'm fine if we don't take the average MAXRSS from multiple runs as long as we maintain the same number of columns in the output. We can then monitor the results to see if we notice problems in action.

palimondo commented:
@lplarson In the discussion on [SR-4667](https://bugs.swift.org/browse/SR-4667) @aschwaighofer explained where does the whole variance in memory usage between runs come from: `Benchmark_O` and co. are adaptive — they run different number of iterations (N) between runs — to get roughly 1s runtime for each test. Coupled with autorelease calling convention for ObjC object you end up with different amounts of garbage between runs. If you specify `num-iterations` they are memory stable (-+ few pages). 

So I'm thinking we need to monitor memory usage separately, in a different run with fixed num-iterations?

lplarson commented:
If we can use a static number of iterations and get reliable results, then I'm all for it. We can potentially speed up benchmarking by not running each test for a second as well.

@aschwaighofer, @gottesmm Am I missing a profound reason to adaptively find iteration counts that run for 1 second apart from convenience to the person adding a test? If it's only convenience, it seems easy enough to create a utility that will find the proper iteration count and set it statically in the test before committing. This would cut down on some noise in the benchmarks.

aschwaighofer commented:
@lplarson I am not the author. I see no profound reason. We will need different counts for O and Onone. Also, it will vary on platforms. But presumably not by a significant factor so if you aim for 0.1 s you will probably be fine on slower platforms.

atrick commented:
@lplarson @aschwaighofer @gottesmm The driver should first determine how many iterations are necessary for a sufficient sample period. 1s is seen as long enough to factor out the timing and harness overhead. The number of iterations varies wildly between revisions of the compiler. It's critically important that a badly behaving benchmark doesn't sabotage a set of runs by taking minutes to complete, and it's important that benchmarks remain measurable after they are dramatically improved. I've never heard of this technique as introducing noise. It's supposed to reduce noise. At any rate, the driver became much more useful after Michael Z (Mr. zperf) implemented this.

atrick commented:
@lplarson @palimondo Oh, I see where this fixed iteration question came from. Measure memory utilization and cpu cycles is a totally different thing. I think it should be a separate run with a separate report.

atrick commented:
One more thing :) I really don't want my benchmark runs to be slowed down by measuring memory usage. It's a great thing to track and helpful to reveal leaks, but not something I want to look at 99% of the time.

palimondo commented:
That was my point. We should kick the MAX_RSS measurement out of the timing measurement and do a separate pass for memory only. Then we don’t need to re-implement whole averaging in `Benchmark_Driver`again.

I think this would be better doable in a separate log file even. I suggest following:
* To get the first baseline: do a single `Benchmark_O` sample run, with verbose setting - extract the computed N (`Measuring with scale ###`) for each test (that’s the part that ensures ~1s runtime - makes sense to maintain it) 
* If baseline exists (i.e. comparing against old memory log file), get Ns for test from there
* Run each test instrumented to extract MAX_RSS with proper fixed N
This is all with `--num-samples=1`, this should take about 5 minutes on my machine.
* Log test name, num-iterations, MAX_RSS
* ????
* PROFIT!!!

This frees `Benchmark_Driver` to completely rely on `Benchmark_O` for normal performance run. It even supports filtering by index number, so we can pass all this in single invocation. Currently the `Benchmark_Driver` does **`numberOfTests * iterations`** separate invocations of `Benchmark_O`.

Do I have a green light to implement this?

atrick commented:
@palimondo That sounds great. The technique of parsing the `scale` from the first (zero-sample) run and reusing it for subsequent runs will also be needed for measuring CPU time across multiple invocations of Benchmark_O (see `--rerun` feature). In the future we'll even want to do that across multiple builds of Benchmark_O that we want to compare.

lplarson commented:
Sounds good to me.

lplarson commented:
Please work with @shahmishal to make sure benchmark pull request testing isn't disrupted.

palimondo commented:
@lplarson Re: Work with @shahmishal. What do you mean, exactly?

palimondo commented:
Wow… 😲

```
$ $BUILD/bin/Benchmark_O --verbose --num-samples=5 99
Verbose
--- CONFIG ---
NumSamples: 5
Verbose: true
IterScale: 1
Tests Filter: ["99"]
Tests to run: DropWhileSequenceLazy, 

--- DATA ---
#,TEST,SAMPLES,MIN(us),MAX(us),MEAN(us),SD(us),MEDIAN(us)
Running DropWhileSequenceLazy for 5 samples.
    Measuring with scale 840.
    Sample 0,304
    Measuring with scale 3601.
    Sample 1,290
    Measuring with scale 2198.
    Sample 2,301
    Measuring with scale 2832.
    Sample 3,298
    Measuring with scale 3596.
    Sample 4,298
99,DropWhileSequenceLazy,5,290,304,298,5,298

Totals,1,290,304,298,0,0
```

Scale (num-iters) varies between samples? How are these then comparable?!?

palimondo commented:
I’d argue that the `scale` computation in `DriverUtils` should be before the `for s in 0..<c.numSamples`.

palimondo commented:
Re: How are these comparable… `samples[s] = elapsed_time / UInt64(scale) / 1000` 💡 🤸‍♂️

lplarson commented:
@palimondo Let @shahmishal know if you need to change something that will break "@swiftci please benchmark"

palimondo commented:
@shahmishal @lplarson Where can I see how the "@swiftci please benchmark” works? Is the GitHub integration visible somewhere? I’d love to know how the CI works in more detail, in order to fix [SR-4675](https://bugs.swift.org/browse/SR-4675) Report detailed build version from Benchmark drivers.

In the final stage, I want to see in the report the unique identification of the branch and commit of the two compared versions. These also need to be in the logs from `Benchmark_Driver`. The baby step landed in #8978 Fix SR-4659 Benchmark logs should be tied to tested tree version, but that depends on the state of working directory at the time of benchmark. This works fine for CI, but is not good enough for local development — ergo SR-4675.
